From: Hendrik Brueckner <brueckner@redhat.com>
Date: Wed, 4 Aug 2010 10:02:54 -0400
Subject: [s390] zfcp: Fail erp after timeout
Message-id: <1280916181-27024-2-git-send-email-brueckner@redhat.com>
Patchwork-id: 27368
O-Subject: [RHEL5.6 PATCH 1/8] [s390x] zfcp: Fail erp after timeout
Bugzilla: 612261
RH-Acked-by: Pete Zaitcev <zaitcev@redhat.com>

Description
-----------
The erp does not recover from temporary interruptions at
the remote port.

The main erp strategy routine does not check for the
TIMEDOUT flag.  Instead, it tries the next erp step as
if the timeout did not happen.

Check for the TIMEDOUT flag and fail the erp action after
a timeout. The fail will trigger appropriate followups.

Bugzilla
--------
BZ 612261
https://bugzilla.redhat.com/show_bug.cgi?id=612261

Upstream status of the patch
----------------------------
http://git.kernel.org/?p=linux/kernel/git/jejb/scsi-misc-2.6.git;a=commit;h=9c785d944e6fa7eef390c799b93e43243505780c

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

diff --git a/drivers/s390/scsi/zfcp_erp.c b/drivers/s390/scsi/zfcp_erp.c
index 8d7c761..a1673a7 100644
--- a/drivers/s390/scsi/zfcp_erp.c
+++ b/drivers/s390/scsi/zfcp_erp.c
@@ -1072,6 +1072,11 @@ zfcp_erp_strategy(struct zfcp_erp_action *erp_action)
 		goto unlock;
 	}
 
+	if (erp_action->status & ZFCP_STATUS_ERP_TIMEDOUT) {
+		retval = ZFCP_ERP_FAILED;
+		goto check_target;
+	}
+
 	/*
 	 * move action to 'running' queue before processing it
 	 * (to avoid a race condition regarding moving the
@@ -1138,6 +1143,7 @@ zfcp_erp_strategy(struct zfcp_erp_action *erp_action)
 	}
 	/* ok, finished action (whatever its result is) */
 
+check_target:
 	/* check for unrecoverable targets */
 	retval = zfcp_erp_strategy_check_target(erp_action, retval);
 
